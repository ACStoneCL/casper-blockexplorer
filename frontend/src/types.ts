export type Account = {
  trimmedAccountHash: string;
  rawAccountHash: string;
  publicKey: string;
  mainPurse: string;
}

export type Block = {
  height: number;
  eraID: number;
  deployCount: number;
  deployHashes?: string[];
  transferHashes?: string[];
  timestamp: number;
  hash: string;
  validatorPublicKey: string;
  parentHash: string;
  stateRootHash?: string;
  proofs?: any[];
};

export type Peer = {
  id: string;
  address: string;
};

export enum DeployStatus {
  Success,
  Failed,
}

export type Deploy = {
  timestamp: number;
  deployHash: string;
  blockHash: string;
  publicKey: string;
  status: DeployStatus;
};
