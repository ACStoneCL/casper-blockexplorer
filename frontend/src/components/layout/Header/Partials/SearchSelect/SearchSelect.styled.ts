import styled from '@emotion/styled';

export const MobileSelectContainer = styled.ul`
  display: flex;
  justify-content: space-between;
  width: 20.5rem;
  margin: 0 auto;
`;

export const MobileSelectButton = styled.button`
  font-size: 0.8rem;
  border-style: none;
  font-weight: 500;
  border-radius: 0.938rem;
  padding: 0.3125rem 0.625rem;
`;

export const SelectWrapper = styled.div`
  .react-select__control {
    width: 11rem;
    box-shadow: none;
    border: none;
  }

  .react-select__value-container {
    height: 3.2rem;
    background-color: #fff;
    padding: 0.7rem 0rem 0.7rem 0rem;
    margin: 0;
    border-radius: 0.375rem 0 0 0.375rem;
    box-shadow: inset 0px 1px 7px rgba(127, 128, 149, 0.3);
    position: relative;
  }

  .react-select__indicators {
    display: block;
    position: absolute;
    top: 0.5rem;
    left: 7.938rem;
    padding: 0;
  }

  .react-select__value-container:hover {
    cursor: pointer;
  }

  .react-select__single-value {
    color: black;
    font-weight: 500;
    font-size: clamp(1rem, 1vw, 1.1rem);
    text-align: left;
    padding-right: 2.3rem;
    padding-left: 1rem;
    padding-bottom: 0.05rem;
    margin: 0 auto;
  }

  .react-select__dropdown-indicator svg {
    color: black;
    width: 25px;
  }

  .react-select__indicator {
    display: block;
    color: black;
    width: 5px;
    padding: 0.5rem 0.4rem 0.5rem 0;
  }

  .react-select__menu-list {
    color: black;
    font-size: clamp(1rem, 1.2vw, 1.4rem);
    padding: 0;
    margin: 0;
    border-radius: 0.375rem;
  }

  .react-select__menu {
    color: white;
    background-color: #fff;
    border-radius: 0.375rem;
    width: 10.9rem;
  }

  .react-select__option:hover,
  .react-select__option:focus {
    color: black;
    transition-property: font-weight, color, 200ms, ease-in;
    background-image: linear-gradient(
      90deg,
      #1c1e90,
      #693590,
      #d81d54,
      #d81e54,
      #fd6b52
    );
    background-size: 100%;
    -webkit-background-clip: text;
    -moz-background-clip: text;
    -webkit-text-fill-color: transparent;
    -moz-text-fill-color: transparent;
    font-weight: 700;
    background-color: transparent;
  }
  .react-select__option {
    background-color: transparent;
  }

  .react-select__option--is-selected {
    background-color: #0721cb;
  }
`;
