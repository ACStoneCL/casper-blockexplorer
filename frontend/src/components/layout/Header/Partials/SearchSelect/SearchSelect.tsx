import React from 'react';
import { Control, Controller } from 'react-hook-form';

import { FormValues, SelectOptions } from '../partials.types';
import { useAppWidth } from '../../../../../hooks';
import {
  MobileSelectButton,
  MobileSelectContainer,
} from './SearchSelect.styled';
import { CustomSelect } from '../CustomSelect';

export interface SearchSelectProps {
  readonly control: Control<FormValues, any>;
  readonly currentFilterOption: string;
  readonly setCurrentFilterOption: React.Dispatch<React.SetStateAction<string>>;
}

export const SearchSelect: React.FC<SearchSelectProps> = ({
  control,
  currentFilterOption,
  setCurrentFilterOption,
}) => {
  const { isMobile } = useAppWidth();

  const selectOptions: SelectOptions[] | null = [
    { value: 'account', label: 'Account' },
    { value: 'deploy', label: 'Deploy' },
    { value: 'block', label: 'Block Hash' },
    { value: 'blockHeight', label: 'Block Height' },
  ];

  return (
    <section data-testid="search-select">
      <Controller
        data-testid="controller"
        control={control}
        render={({ field: { onChange, value, name } }) => {
          const currentSelection = selectOptions.find(
            option => option.value === value,
          );

          const handleSelectChange = (selectedOption: SelectOptions | null) => {
            onChange(selectedOption?.value);
            setCurrentFilterOption(selectedOption?.value!);
          };

          return !isMobile ? (
            <CustomSelect
              defaultValue={selectOptions[0]}
              currentSelection={currentSelection}
              name={name}
              options={selectOptions}
              onChange={handleSelectChange}
            />
          ) : (
            <MobileSelectContainer>
              {selectOptions.map(option => {
                const handleClick = () => {
                  onChange(option.value);
                  setCurrentFilterOption(option.value);
                };

                return (
                  <li key={option.value}>
                    <MobileSelectButton
                      onClick={handleClick}
                      style={{
                        backgroundColor:
                          currentFilterOption === option.value
                            ? '#0325d1'
                            : '#F1F1F4',
                        color:
                          currentFilterOption === option.value
                            ? '#fff'
                            : '#000',
                      }}
                      type="button">
                      {option.label.includes('Block')
                        ? option.label.replace('Block', 'Blk')
                        : option.label}
                    </MobileSelectButton>
                  </li>
                );
              })}
            </MobileSelectContainer>
          );
        }}
        name="filterOptions"
        rules={{
          required: true,
        }}
      />
    </section>
  );
};
