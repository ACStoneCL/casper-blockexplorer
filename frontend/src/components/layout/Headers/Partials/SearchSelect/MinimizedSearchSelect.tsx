import React from 'react';
import { Control, Controller } from 'react-hook-form';

import { FormValues, SelectOptions } from '../partials.types';
import { useAppWidth } from '../../../../../hooks';
import {
  MobileSelectButton,
  MobileSelectContainer,
} from './SearchSelect.styled';
import { MinimizedCustomSelect } from '../CustomSelect';

export interface SearchSelectProps {
  readonly control: Control<FormValues, any>;
  readonly currentFilterOption: string;
  readonly setCurrentFilterOption: React.Dispatch<React.SetStateAction<string>>;
}

export const MinimizedSearchSelect: React.FC<SearchSelectProps> = ({
  control,
  currentFilterOption,
  setCurrentFilterOption,
}) => {
  const { isMobile } = useAppWidth();

  const selectOptions: SelectOptions[] | null = [
    { value: 'account', label: 'Account' },
    { value: 'deploy', label: 'Deploy' },
    { value: 'block', label: 'Block Hash' },
    { value: 'blockHeight', label: 'Block Height' },
  ];

  return (
    <section data-testid="search-select">
      <Controller
        data-testid="controller"
        control={control}
        render={({ field: { onChange, value, name } }) => {
          const currentSelection = selectOptions.find(
            option => option.value === value,
          );

          const handleSelectChange = (selectedOption: SelectOptions | null) => {
            onChange(selectedOption?.value);
            setCurrentFilterOption(selectedOption?.value!);
          };

          return !isMobile ? (
            <MinimizedCustomSelect
              defaultValue={selectOptions[0]}
              currentSelection={currentSelection}
              name={name}
              options={selectOptions}
              onChange={handleSelectChange}
            />
          ) : (
            <MobileSelectContainer>
              {selectOptions.map((option, index) => {
                const handleClick = () => {
                  onChange(option.value);
                  setCurrentFilterOption(option.value);
                };

                return (
                  <li key={option.value} data-testid={`button-${index + 1}`}>
                    <MobileSelectButton
                      onClick={handleClick}
                      style={{
                        backgroundColor:
                          currentFilterOption === option.value
                            ? '#0325d1'
                            : '#F1F1F4',
                        color:
                          currentFilterOption === option.value
                            ? '#fff'
                            : '#000',
                      }}
                      type="button">
                      {option.label.includes('Block')
                        ? option.label.replace('Block', 'Blk')
                        : option.label}
                    </MobileSelectButton>
                  </li>
                );
              })}
            </MobileSelectContainer>
          );
        }}
        name="filterOptions"
        rules={{
          required: true,
        }}
      />
    </section>
  );
};
