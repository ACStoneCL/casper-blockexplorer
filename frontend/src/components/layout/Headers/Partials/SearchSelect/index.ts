export { SearchSelect } from './SearchSelect';
export { MinimizedSearchSelect } from './MinimizedSearchSelect';
