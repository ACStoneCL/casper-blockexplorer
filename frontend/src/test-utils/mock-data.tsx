import { Block } from '../api';

export const mockBlock: Block = {
  height: 1,
  eraID: 1,
  deployCount: 1,
  deployHashes: ['deploy-hash'],
  transferHashes: [],
  timestamp: 123456789,
  readableTimestamp: 'Dec 25th, 1975',
  timeSince: 'Forever Ago..',
  hash: 'block-hash',
  validatorPublicKey: 'validator-public-key',
  parentHash: 'parent-hash',
  stateRootHash: 'state-root-hash',
  rawBlock: '{}',
};

export const createMockBlocks = (numberOfBlocks: number = 10) => {
  const blocks = new Array(numberOfBlocks).fill(mockBlock) as Block[];

  return blocks.map((block, index) => {
    return { ...block, hash: `${block.hash}-${index}` };
  });
};
