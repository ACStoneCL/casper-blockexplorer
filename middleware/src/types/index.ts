export * from "./nodeStatusResponse";
export * as OnChain from "./on-chain";
export type Sort = "DESC" | "ASC";
