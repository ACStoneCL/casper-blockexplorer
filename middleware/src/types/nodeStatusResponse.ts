export interface StatusResponse {
  api_version: string;
  chainspec_name: string;
  starting_state_root_hash: string;
  peers: Peer[];
  last_added_block_info: LastAddedBlockInfo;
  our_public_signing_key: string;
  round_length: null;
  next_upgrade: null;
  build_version: string;
  uptime: string;
}

export interface LastAddedBlockInfo {
  hash: string;
  timestamp: Date;
  era_id: number;
  height: number;
  state_root_hash: string;
  creator: string;
}

export interface Peer {
  node_id: string;
  address: string;
}
