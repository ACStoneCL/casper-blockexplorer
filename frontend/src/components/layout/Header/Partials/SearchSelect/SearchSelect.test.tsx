import React from 'react';
import { useForm } from 'react-hook-form';
import { Context as ResponsiveContext } from 'react-responsive';
import { render, renderHook } from '../../../../../test-utils';
import { SearchSelect } from './SearchSelect';
import { FormValues } from '../partials.types';

window.resizeTo = function resizeTo(width, height): void {
  Object.assign(this, {
    innerWidth: width,
    innerHeight: height,
    outerWidth: width,
    outerHeight: height,
  }).dispatchEvent(new this.Event('resize'));
};

describe('SearchSelect', () => {
  it('should render without error', () => {
    const { result } = renderHook(useForm<FormValues>);
    const mockCurrentFilterOption = 'current-option';
    const { getByTestId } = render(
      <SearchSelect
        control={result.current.control}
        currentFilterOption={mockCurrentFilterOption}
        setCurrentFilterOption={jest.fn()}
      />,
    );

    const searchSelectWrapper = getByTestId('search-select');

    expect(searchSelectWrapper).toBeInTheDocument();
  });

  it('should not display react-select component if screen width is less than 1023px', () => {
    const { result } = renderHook(useForm<FormValues>);

    const mockCurrentFilterOption = 'current option';
    const { container: mobile } = render(
      // eslint-disable-next-line react/jsx-no-constructed-context-values
      <ResponsiveContext.Provider value={{ width: 1022 }}>
        <SearchSelect
          control={result.current.control}
          currentFilterOption={mockCurrentFilterOption}
          setCurrentFilterOption={jest.fn()}
        />
      </ResponsiveContext.Provider>,
    );

    expect(mobile).toMatchSnapshot();
  });
});
