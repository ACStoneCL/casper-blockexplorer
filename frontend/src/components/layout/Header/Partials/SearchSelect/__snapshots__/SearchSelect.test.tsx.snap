// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSelect should not display react-select component if screen width is less than 1023px 1`] = `
<div>
  <section
    data-testid="search-select"
  >
    <ul
      class="css-1mwiq5w"
    >
      <li>
        <button
          class="css-1u867rd"
          style="background-color: rgb(241, 241, 244); color: rgb(0, 0, 0);"
          type="button"
        >
          Account
        </button>
      </li>
      <li>
        <button
          class="css-1u867rd"
          style="background-color: rgb(241, 241, 244); color: rgb(0, 0, 0);"
          type="button"
        >
          Deploy
        </button>
      </li>
      <li>
        <button
          class="css-1u867rd"
          style="background-color: rgb(241, 241, 244); color: rgb(0, 0, 0);"
          type="button"
        >
          Blk Hash
        </button>
      </li>
      <li>
        <button
          class="css-1u867rd"
          style="background-color: rgb(241, 241, 244); color: rgb(0, 0, 0);"
          type="button"
        >
          Blk Height
        </button>
      </li>
    </ul>
  </section>
</div>
`;
